File: 21. ReverseWordsInAString.cpp
#include <iostream>
#include <string>
#include <vector>
#include <sstream>
using namespace std;

string reverseWords(string s) {
    istringstream iss(s);
    vector<string> words;
    string w;
    while (iss >> w) words.push_back(w);
    string result;
    for (int i = words.size() - 1; i >= 0; i--) {
        result += words[i];
        if (i > 0) result += ' ';
    }
    return result;
}

int main() {
    cout << reverseWords("the sky is blue") << "\n";
    cout << reverseWords("  hello world  ") << "\n";
    cout << reverseWords("a good   example") << "\n";

    return 0;
}

File: 21. ReverseWordsInAString.java
public class ReverseWordsInAString {

    public static String reverseWords(String s) {
        String[] words = s.trim().split("\\s+");
        StringBuilder sb = new StringBuilder();
        for (int i = words.length - 1; i >= 0; i--) {
            sb.append(words[i]);
            if (i > 0) sb.append(' ');
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        System.out.println(reverseWords("the sky is blue"));       // "blue is sky the"
        System.out.println(reverseWords("  hello world  "));       // "world hello"
        System.out.println(reverseWords("a good   example"));      // "example good a"
    }
}
